MAINT: fix some small array API support issues

rgommers opened the pull request:
These surfaced with the more strict reference implementation in https://github.com/numpy/numpy/pull/25370 and with the addition of a `__array_namespace__` method to `numpy.ndarray` for NumPy 2.0 (which caused the `is_numpy` utility function to be wrong). Even if the `is_numpy` issue is also handled by a change in array-api-compat, this still seems like a logical change to make; if `x.__name__` is `'numpy'` then `is_numpy(x)` should return True.

Further changes:
- improved tracebacks for test functions in `_lib/_array_api.py`
- update the `array-ap-compat` submodule to 1.4.1, in order to support NumPy `main`/2.0

rgommers committed:
MAINT: fix some small array API support issues

These surfaced with the more strict reference implementation in
https://github.com/numpy/numpy/pull/25370 and with the addition
of a `__array_namespace__` method to `numpy.ndarray` for NumPy 2.0
(which caused the `is_numpy` utility function to be wrong).
Even if the `is_numpy` issue is also handled by a change in
array-api-compat, this still seems like a logical change to make;
if `x.__name__` is `'numpy'` then `is_numpy(x)` should return True.

Regarding the tightening up of bool dtype handling in `_lib.lazywhere`:
plain `bool` is technically not guaranteed to work, and the stricter
checks in the reference implementation will start rejecting it.

[skip cirrus] [skip circle]

rgommers commented:
Hmm, the test failure on 32-bit is real. It looks like `np.take` handles 64-bit input on 32-bit Python less robustly than the indexing implementation does:
```
/scipy/test/lib/python3.9/site-packages/numpy/core/fromnumeric.py:43: in _wrapit
    result = getattr(asarray(obj), method)(*args, **kwds)
E   TypeError: Cannot cast array data from dtype('int64') to dtype('int32') according to the rule 'safe'
        args       = (array([194, 195], dtype=int64),)
        kwds       = {'axis': 0, 'mode': 'raise', 'out': None}
        method     = 'take'
```
full traceback:

<details>

```
 ________________ TestKMean.test_kmeans_and_kmeans2_random_seed _________________
/scipy/test/lib/python3.9/site-packages/numpy/core/fromnumeric.py:57: in _wrapfunc
    return bound(*args, **kwds)
E   TypeError: Cannot cast array data from dtype('int64') to dtype('int32') according to the rule 'safe'
        args       = (array([194, 195], dtype=int64),)
        bound      = <built-in method take of numpy.ndarray object at 0xec0303b0>
        kwds       = {'axis': 0, 'mode': 'raise', 'out': None}
        method     = 'take'
        obj        = array([[-2.2 ,  1.17],
       [-1.63,  1.69],
       [-2.04,  4.38],
       [-3.09,  0.95],
       [-1.7 ,  4.79],
   ...1],
       [-2.88,  1.56],
       [-1.63,  1.46],
       [-1.41,  0.68],
       [-1.96,  2.76],
       [-1.61,  2.11]])

During handling of the above exception, another exception occurred:
scipy/cluster/tests/test_vq.py:408: in test_kmeans_and_kmeans2_random_seed
    res1, _ = kmeans(TESTDATA_2D, 2, seed=seed)
        _          = array([1, 1, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 1, 1,
       1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 1, 1, 0,...1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1,
       1, 1, 1, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1,
       0, 1])
        minit      = '++'
        res1       = array([[-1.95266667,  0.898     ],
       [-3.153375  ,  3.3945    ]])
        res2       = array([[-3.153375  ,  3.3945    ],
       [-1.95266667,  0.898     ]])
        seed       = Generator(PCG64) at 0xF4C9B0D8
        seed_list  = [1234, RandomState(MT19937) at 0xF4BC63E8, Generator(PCG64) at 0xF4C9B0D8]
        self       = <scipy.cluster.tests.test_vq.TestKMean object at 0xec03b9e8>
scipy/cluster/vq.py:491: in kmeans
    guess = _kpoints(obs, k, rng, xp)
        best_dist  = inf
        check_finite = True
        guess      = array(2)
        i          = 0
        iter       = 20
        k          = 2
        k_or_guess = 2
        obs        = array([[-2.2 ,  1.17],
       [-1.63,  1.69],
       [-2.04,  4.38],
       [-3.09,  0.95],
       [-1.7 ,  4.79],
   ...1],
       [-2.88,  1.56],
       [-1.63,  1.46],
       [-1.41,  0.68],
       [-1.96,  2.76],
       [-1.61,  2.11]])
        rng        = Generator(PCG64) at 0xF4C9B0D8
        seed       = Generator(PCG64) at 0xF4C9B0D8
        thresh     = 1e-05
        xp         = <module 'scipy._lib.array_api_compat.numpy' from '/scipy/build-install/lib/python3.9/site-packages/scipy/_lib/array_api_compat/numpy/__init__.py'>
scipy/cluster/vq.py:520: in _kpoints
    return xp.take(data, idx, axis=0)
        data       = array([[-2.2 ,  1.17],
       [-1.63,  1.69],
       [-2.04,  4.38],
       [-3.09,  0.95],
       [-1.7 ,  4.79],
   ...1],
       [-2.88,  1.56],
       [-1.63,  1.46],
       [-1.41,  0.68],
       [-1.96,  2.76],
       [-1.61,  2.11]])
        idx        = array([194, 195], dtype=int64)
        k          = 2
        rng        = Generator(PCG64) at 0xF4C9B0D8
        xp         = <module 'scipy._lib.array_api_compat.numpy' from '/scipy/build-install/lib/python3.9/site-packages/scipy/_lib/array_api_compat/numpy/__init__.py'>
<__array_function__ internals>:180: in take
    ???
        args       = (array([[-2.2 ,  1.17],
       [-1.63,  1.69],
       [-2.04,  4.38],
       [-3.09,  0.95],
       [-1.7 ,  4.79],
  ... [-1.63,  1.46],
       [-1.41,  0.68],
       [-1.96,  2.76],
       [-1.61,  2.11]]), array([194, 195], dtype=int64))
        dispatcher = <function _take_dispatcher at 0xeed65a00>
        implementation = <function take at 0xeed65a90>
        kwargs     = {'axis': 0}
        public_api = <function take at 0xeed65ad8>
        relevant_args = (array([[-2.2 ,  1.17],
       [-1.63,  1.69],
       [-2.04,  4.38],
       [-3.09,  0.95],
       [-1.7 ,  4.79],
  ...    [-2.88,  1.56],
       [-1.63,  1.46],
       [-1.41,  0.68],
       [-1.96,  2.76],
       [-1.61,  2.11]]), None)
/scipy/test/lib/python3.9/site-packages/numpy/core/fromnumeric.py:190: in take
    return _wrapfunc(a, 'take', indices, axis=axis, out=out, mode=mode)
        a          = array([[-2.2 ,  1.17],
       [-1.63,  1.69],
       [-2.04,  4.38],
       [-3.09,  0.95],
       [-1.7 ,  4.79],
   ...1],
       [-2.88,  1.56],
       [-1.63,  1.46],
       [-1.41,  0.68],
       [-1.96,  2.76],
       [-1.61,  2.11]])
        axis       = 0
        indices    = array([194, 195], dtype=int64)
        mode       = 'raise'
        out        = None
/scipy/test/lib/python3.9/site-packages/numpy/core/fromnumeric.py:66: in _wrapfunc
    return _wrapit(obj, method, *args, **kwds)
        args       = (array([194, 195], dtype=int64),)
        bound      = <built-in method take of numpy.ndarray object at 0xec0303b0>
        kwds       = {'axis': 0, 'mode': 'raise', 'out': None}
        method     = 'take'
        obj        = array([[-2.2 ,  1.17],
       [-1.63,  1.69],
       [-2.04,  4.38],
       [-3.09,  0.95],
       [-1.7 ,  4.79],
   ...1],
       [-2.88,  1.56],
       [-1.63,  1.46],
       [-1.41,  0.68],
       [-1.96,  2.76],
       [-1.61,  2.11]])
/scipy/test/lib/python3.9/site-packages/numpy/core/fromnumeric.py:43: in _wrapit
    result = getattr(asarray(obj), method)(*args, **kwds)
E   TypeError: Cannot cast array data from dtype('int64') to dtype('int32') according to the rule 'safe'
        args       = (array([194, 195], dtype=int64),)
        kwds       = {'axis': 0, 'mode': 'raise', 'out': None}
        method     = 'take'
        obj        = array([[-2.2 ,  1.17],
       [-1.63,  1.69],
       [-2.04,  4.38],
       [-3.09,  0.95],
       [-1.7 ,  4.79],
   ...1],
       [-2.88,  1.56],
       [-1.63,  1.46],
       [-1.41,  0.68],
       [-1.96,  2.76],
       [-1.61,  2.11]])
        wrap       = <built-in method __array_wrap__ of numpy.ndarray object at 0xec0303b0>
=========================== short test summary info ============================
FAILED scipy/cluster/tests/test_vq.py::TestKMean::test_kmeans_and_kmeans2_random_seed
```

</details>

lucascolley commented on scipy/_lib/_array_api.py:
```diff
@@ -196,15 +196,15 @@ def copy(x, *, xp=None):
-    return xp.__name__ == 'scipy._lib.array_api_compat.numpy'
+    return xp.__name__ in ('numpy', 'scipy._lib.array_api_compat.numpy')
 
 
 def is_cupy(xp):
-    return xp.__name__ == 'scipy._lib.array_api_compat.cupy'
+    return xp.__name__ in ('cupy', 'scipy._lib.array_api_compat.cupy')
 
 
 def is_torch(xp):
-    return xp.__name__ == 'scipy._lib.array_api_compat.torch'
+    return xp.__name__ in ('torch', 'scipy._lib.array_api_compat.torch')
```
These were deliberate at the time, since NumPy (lacking `__array_namespace__`) would always return `array_api_compat.numpy` from `array_namespace`. But yes, I think it's good to change to this now (it used to be `'numpy' in xp.__name__`).

lucascolley commented on scipy/cluster/hierarchy.py:
```diff
@@ -163,7 +163,7 @@ def _warning(s):
     # numpy.array_api is strict about not allowing `int()` on a float array.
     # That's typically not needed, here it is - so explicitly convert
-    return int(xp.astype(arr, xp.int64))
+    return int(xp.astype(xp.asarray(arr), xp.int64))
```
```suggestion
    return int(xp.asarray(arr, dtype=xp.int64))
```

lucascolley replied:
Also, why is this change needed? Does `xp` not come from `array_namespace(arr)`?

rgommers replied:
I'll verify, but IIRC `arr` can be either a 1-D array or a scalar, and the latter didn't work.

rgommers replied:
Actually, it was a scalar only because we were bypassing `array_api_compat`; now that that's fixed this change is no longer needed it looks like.

rgommers replied:
The problem stems from indexing; arrays are converted correctly but then `arr = row[0]` may produce a scalar again - and `int_floor` is called from multiple places.

rgommers replied:
No, it doesn't quite, because (from the [`asarray` docs](https://data-apis.org/array-api/latest/API_specification/generated/array_api.asarray.html)):

<img width="502" alt="image" src="https://github.com/scipy/scipy/assets/98330/d60c345a-f3b3-4bd6-ab48-44af34a0eddd">

For unsafe casting from float to int, `astype` should be correct here.

lucascolley replied:
Ah yes, I remember that pattern now that you mention it!

rgommers committed:
TST: array-api: improve tracebacks for testing functions

`__tracebackhide__ = True` should be used for all testing functions,
just like it is in `numpy.testing`. This makes error messages much
more readable.

rgommers commented:
Converting this to draft until I can include an update of `array_api_compat` that supports numpy 2.0 - see https://github.com/data-apis/array-api-compat/issues/77.

rgommers committed:
MAINT: update array-api-compat to version 1.4.1

This version has support for NumPy 2.0

rgommers commented:
This now includes an update of `array-api-compat` to version `1.4.1`, and should be good to go.

lucascolley reviewed (approved):
LGTM

rgommers commented:
Thanks for the review @lucascolley. I'll go ahead and merge this, to unblock https://github.com/numpy/numpy/pull/25370.
